File: Leetcode/267.py

#266. Palindrome Permutation II

import copy
class stack:
    def __init__(self):
        self.body = []

    def push(self, data):
        self.body.append(data)

    def pop(self):
        if (len(self.body) == 0):
            return None
        else:
            return self.body.pop()

    def isEmpty(self):
        return True if len(self.body) == 0 else False

def getPerm(freq):
    result = []
    count = 0
    for k in freq.keys(): count += freq[k]
    s = stack()
    s.push(("",{}))
    while(not s.isEmpty()):
        sp, p = s.pop()
        if (len(sp) == count):
            result.append(sp)
            continue
        for k in freq:
            f = freq[k]
            if (k in p): f -= p[k]
            if (f <= 0): continue
            pp = copy.copy(p)
            if (not k in pp): pp[k] = 1
            else: pp[k] += 1
            s.push((sp+k,pp))
            
    return result       

def allPalPerm(ss):
    freq = {}
    for i in range(len(ss)):
        s = ss[i]
        if (not s in freq):
            freq[s] = 1
        else:
            freq[s] += 1
        
    odd = ""
    for k in freq.keys():
        f = freq[k]
        if ((f % 2) == 1):
            if (odd == ""):
                odd = k
            else:
                return []

    if (odd != ""): del freq[odd]
    for k in freq.keys():
        freq[k] //= 2

    #print(freq)
    ssl = getPerm(freq)
    result = [ sl+odd+sl[::-1] for sl in ssl ]     
    return result

sample = "code"
print("{0}=>{1}".format(sample, allPalPerm(sample)))
sample = "aab"
print("{0}=>{1}".format(sample, allPalPerm(sample)))
sample = "carerac"
print("{0}=>{1}".format(sample, allPalPerm(sample)))
sample = "ccaaeaacc"
print("{0}=>{1}".format(sample, allPalPerm(sample)))

File: Leetcode/267.cpp


// 266. Palindrome Permutation II
#include <iostream>
#include <stack>
#include <string>
#include <vector>
#include <map>

using namespace std;
typedef map<char, int> charfreq;
typedef vector<string> stringarray;

void getFreq(string ss, charfreq& freq)
{
    for(int i=0; i < ss.length() ; i++)
    {
        char c = ss[i];
	if (freq.find(c) == freq.end())
	    freq[c] = 0;    
        freq[c] += 1;
    }
}

void getPerm(charfreq freq, stringarray& result)
{
    stack<string> s;
    int count = 0;
    for (charfreq::iterator iter = freq.begin();
         iter != freq.end(); iter++)
        count += iter->second;

    //cout << "sum count = " << count << "\n";
    s.push("");
    while(!s.empty())
    {
        string sp = s.top();
	s.pop();
	if (sp.length() == count)
	{
	    result.push_back(sp);
	    continue;
	}
	 
	charfreq p;
	getFreq(sp, p);

        for (charfreq::iterator iter = freq.begin();
             iter != freq.end(); iter++)
	{
            int f = iter->second;
	    if (p.find(iter->first) != p.end())
		f -= p[iter->first];
            if (f > 0) s.push(sp + iter->first);
	}
    }
}

void allPalPerm(string ss, stringarray& result)
{
    charfreq freq;
    getFreq(ss, freq);

    string odd = "";
    for (charfreq::iterator iter = freq.begin();
         iter != freq.end(); iter++)
    {
	int f = iter->second;
	if ((f % 2) == 1)
	{
            if (odd == "") odd = iter->first; 
            else return;
	}
    }

    if (odd != "") freq.erase(freq.find(odd[0]));
    for (charfreq::iterator iter = freq.begin();
         iter != freq.end(); iter++)
        iter->second /= 2;

    stringarray ssl;
    getPerm(freq, ssl);
    for (stringarray::iterator iter = ssl.begin();
         iter != ssl.end(); iter++)
    {
	string rev(iter->rbegin(), iter->rend());    
        string sl = *iter + odd + rev;
	result.push_back(sl);	
    }
}

void test(string ss)
{
    stringarray result;	
    allPalPerm(ss, result);
    cout << "\"" << ss << "\" total = " << result.size() << "\n";
    for (stringarray::iterator iter = result.begin();
         iter != result.end(); iter++)
        cout << *iter << endl; 
}

void test2()
{
    charfreq freq;
    stringarray result;

    freq['a'] = 1;
    freq['b'] = 2;
    freq['c'] = 3;
    getPerm(freq, result);
    cout << "total = " << result.size() << "\n";
    for (stringarray::iterator iter = result.begin();
         iter != result.end(); iter++)
        cout << *iter << endl; 
}

int main()
{
#if 1 
    test("code");
    test("aab");
    test("carerac");
    test("ccaaeaacc");
#else
    test2();
#endif

    return 0;
}
